lib/streamlit/metrics_util.py

```diff
@@ -0,0 +1,77 @@
+import os
+import platform
+import subprocess
+import threading
+import uuid
+
+from typing import Optional
+
+from streamlit import file_util
+
+def _get_machine_id_v1():
+    """Generate the prior machine ID that we used as a unique identifier for a user while tracking
+    metrics in Segment.  This ID is broken in different ways in some Linux distros and Docker images.
+    - sometimes our machine ID is just a hash of '', which means many machines map to the same ID
+    - sometimes our machine ID is a hash of the same string, when running in a Docker container
+    - we run a sudo command, which is weird and bad in all sorts of ways
+    """
+    if (
+        platform.system() == "Linux"
+        and os.path.isfile("/etc/machine-id") == False
+        and os.path.isfile("/var/lib/dbus/machine-id") == False
+    ):
+        subprocess.run(["sudo", "dbus-uuidgen", "--ensure"])
+
+    machine_id = str(uuid.getnode())
+    if os.path.isfile("/etc/machine-id"):
+        with open("/etc/machine-id", "r") as f:
+            machine_id = f.read()
+
+    elif os.path.isfile("/var/lib/dbus/machine-id"):
+        with open("/var/lib/dbus/machine-id", "r") as f:
+            machine_id = f.read()
+
+    return machine_id
+
+def _get_machine_id_v2():
+    """Generate the new machine ID that we'll as a unique identifier for a user while tracking
+    metrics in Segment. Instead of relying on a hardware address in the container or host we'll
+    generate a UUID and store it in the Streamlit hidden folder."""
+    filepath = file_util.get_streamlit_file_path(".stable_random_id")
+    stable_id = None
+
+    if os.path.exists(filepath):
+        with file_util.streamlit_read(filepath) as input:
```

Comment:
Actually, your code is correct, but only because `streamlit_read` and `streamlit_write` are broken. IIUC they're supposed to read/write the provided file from the Streamlit folder. But because of a quirk in how `os.path.join` works, they're not actually doing that. This was broken a long time ago, in PR https://github.com/streamlit/streamlit/pull/227

We should fix this (in a separate PR), but that'll be fairly hairy as we need to double check that everything that depends on `streamlit_read`/`streamlit_write` works correctly in a bunch of different scenarios.

Reply:
@tconkling FYI 👆